recipe/2023-11-support-insecure-registries.md

```diff
@@ -0,0 +1,185 @@
+  @doc("The Bicep template kind.")
+  templateKind: "bicep";
+
++  @doc("Allows an insecure connection to a Bicep registry without doing an SSL check. Used commonly for locally-hosted registries. Defaults to false (require SSL).")
++  insecureHttp?: boolean;
+}
+
+@doc("The properties of a Recipe linked to an Environment.")
+model RecipeGetMetadataResponse {
+  @doc("The format of the template provided by the recipe. Allowed values: bicep, terraform.")
+  templateKind: string;
+
+  @doc("The path to the template provided by the recipe. Currently only link to Azure Container Registry is supported.")
+  templatePath: string;
+
+  @doc("The version of the template to deploy. For Terraform recipes using a module registry this is required, but must be omitted for other module sources.")
+  templateVersion?: string;
+
+  @doc("The key/value parameters to pass to the recipe template at deployment.")
+  parameters: {};
+
++  @doc("Allows an insecure connection to a Bicep registry without doing an SSL check. Used commonly for locally-hosted registries. Defaults to false (require SSL).")
++  insecureHttp?: boolean;
+}
+```
+
+***CLI changes***
+
+Add insecure-http flag to allow insecure connection to a Bicep registry without doing an SSL check.
+#### Bicep Publish
+
+```
+rad bicep publish --file ./redis-test.bicep --target br:localhost:5000/myregistry/redis-test:v1 --insecure-http
+```
+#### Recipe Register 
+
+```
+rad recipe register cosmosdb --template-kind bicep --template-path br:localhost:5000/myregistry/redis-test:v1 --resource-type Applications.Datastores/mongoDatabases --insecure-http
+```
+
+#### Recipe List/Show
+Add a new column to recipe list/show output table.
+
+| NAME        | TYPE           | TEMPLATE KIND  | TEMPLATE VERSION | TEMPLATE | INSECURE-HTTP|
+| :--: |:--:| :--:| :--: | :--:| :--:|
+| cosmosdb      | Applications.Datastores/mongoDatabases | bicep | | localhost:5000.io/myregistry/redis-test:v1| true| 
+
+***Bicep Changes***
+
+#### Registering recipe through bicep.
+
+Add `insecureHttp` property to use insecure registry.
+```diff
+import radius as radius
+
+@description('Specifies the location for resources.')
+param location string = 'global'
+
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'corerp-resources-recipe-env'
+  location: location
+  properties: {
+    compute: {
+      kind: 'kubernetes'
+      resourceId: 'self'
+      namespace: 'corerp-resources-recipe-env'
+    }
+    recipes: {
+      'Applications.Datastores/mongoDatabases':{
+        recipe1: {
+          templateKind: 'bicep'
+          templatePath: 'ghcr.io/testpublicrecipe/bicep/modules/mongodatabases:v1' 
+        }
+        recipe2: {
+          templateKind: 'bicep'
+          templatePath: 'localhost:5000/myregistry/mongo-test:v1' 
++          insecureHttp: true
+        }
+      }
+    }
+  }
+}
+```
+
+## Alternatives considered
+
+#### Automatically identify the insecure registry.
```

Comment:
I still want to explore if we can special-case `localhost` so `insecure-http` is set to true automatically. The user can still set it to false manually for the edge case where they roll their own SSL certificate locally. And it will allow the publish and register commands to work be default the first time instead of getting an error message.

Reply:
Especially because a user won't know there's a problem for Recipe registration until a Recipe deployment is attempted. Having it "just work" without waiting until a deployment to see the error message is a much better user experience.

Reply:
For context the oras client special cases localhost to use insecure http.

Reply:
> I still want to explore if we can special-case `localhost` so `insecure-http` is set to true automatically. The user can still set it to false manually for the edge case where they roll their own SSL certificate locally. And it will allow the publish and register commands to work be default the first time instead of getting an error message.

I want to argue about this 😆 because I don't think it's useful.

For locally-hosted registries the pull and push URLs are asymmetric. Pushes use localhost, and pulls **DO NOT**. Users still deal with the same level of complexity because Kubernetes is involved. I don't think the heuristic helps. 

Our usecase is different from ORAS' use case. I'd rather be explict, and I don't think this detail is worth explaining to users, it feels like noob-trap.

There's also no such thing as setting a command-line **flag** to false, so it requires a different design. Security should be opt-out rather than opt-in.